### DISCUSSION TITLE
Type for matrix multiplication

### DISCUSSION THREAD

**Discussion by philipJFpfahl** [idaholab/moose#21752] [Q&A Getting Started] [answered]
Hello,
I am new to using moose, but I hope this question is not too trivial. 
I would like to weigh a variable with a pre-calculated matrix and calculate a new variable with that.
$a=Wb$
where $a$ and $b$ would be $N$ dimensional (for each mesh point) and $W$ would be $N \times N$. 
this is part of a coupled equation where
$\dot{p} = p + a$
So far so easy.

Now I wanted to implement this by defining a material "W" and then calling this material later to be consumed by a Kernel doing the matrix multiplication.
First: 
It seems to me, that such an operation should be rather common. Is there already a module doing that?
if not: 
How should I implement the matrix W? 
A Material object seems like the right choice, but I only want to initialize the matrix once. So maybe it's better to write a User object doing that?

Thanks in advance 

``

**Comment by GiudGiud** [comment, marked as answer]
Hello

You may use a RankTwoTensor to store a matrix.
We do have material properties of type rank two tensor. 
See this example
https://mooseframework.inl.gov/source/materials/ADGenericConstantRankTwoTensor.html
Guillaume

**Reply by philipJFpfahl** [reply]
Hello,

thank you for the fast reply. That is, what I was looking for.
So I should make a material, that produces the $W$ matrix and then a kernel that consumes it?
And is it possible to execute the material only at the beginning of my calculations? I tried "execute_on = INITIAL" but that does not seem to work for materials.
I am asking because I am sure what the "best" way to do that is.
Philip

**Reply by GiudGiud** [reply]
Hello

Yes for the material creating / computing it and the kernel consuming / using the value.

I'm afraid not, we dont really support execution schedules on regular materials (only on functor materials, but we havent added the tensor overloads for these). What you could do is make it depend on copies of the variables, then if those copies are never updated the material properties would be constant. 

Guillaume

**Reply by philipJFpfahl** [reply]
Hello,

so the material object is called, whenever the kernel consumes it? That seems like a substantial computational burden if the material stays constant all the time, especially with a larger mesh. I am also not interested in the shape function of the $W$ matrix (since I already know it).
Is it better to store the matrix as a <real> vector?
If yes, how would I generate such a vector?
I could then use it to make the matrix multiplication in the Kernel as a loop.
For $N$ mesh points it would be something like this

for(int i =0; i < $N \times N$; i++)
{
a[_qp] = W[_qp * $N \times N$ + i] b[i]
}

I am wondering why no one uses this kind of operation in MOOSE. Is there a better way to approach this problem?
Anyway thanks for your help this far.

Philip

**Comment by GiudGiud** [comment]
If it s constant spatially then you may use the constant_on option to say that it s constant by block and that will make it only be computed once for every element / qp in the block 

You may store it as a vector if you prefer, please have a look at Vector materials instead of tensors then

**Reply by philipJFpfahl** [reply]
Hey, 

So I reviewed the problem. I think there was a misunderstanding from my side of how MOOSE handles indexing.
When doing the finite element method you have nodes and elements. The number of nodes is equal to the number of points in your mesh. In MOOSE you integrate over these elements. The _qp index indicates the point in the element on which the integral is supported.

What I need to do has nothing to do with the integral. On each time step, I want to adjust the value of each node point. The adjustment depends on the matrix $W$ and all the other nodal points. In the sense, that the jth node is calculated like this
$$a_j = \sum_{k = 1}^N W_{j,k} b_k$$
with N nodal points. 
In a more physical sense, that means, that every point of $a$ depends on all the other points in $b$ multiplied by a weight.

I am sorry for my misleading question, I didn't know better. Is there a way in MOOSE to solve these kinds of problems?

Thanks again,
Philip

**Reply by GiudGiud** [reply]
I think you can do that in MOOSE, though I think you'll have to code it
Have `a` be an AuxVariable, and the kernel that does this operation an AuxKernel. 

Then you ll want to access 
- a at _qp
- b at k, which will be the index in a loop from 0 to Nqp
- W at the combined index

W can still be a vector or a rank two tensor, as you prefer. If it s constant, GenericConstantRankTwoTensor is still a good choice

**Reply by philipJFpfahl** [reply]
Hey 
I had to try that. The problem remains. I can not access the full value of my Variable b in my Kernel. I checked that it is nodal. I suppose I just struggle with the indexing of $b$ in the Kernel.

I initialize it with
  const VariableValue & _b;

and add it to the parameters 
  params.addRequiredCoupledVar("b", "This is text");
  _b(coupledValue("b"))

in the for loop, I try to index it as _b[k], but it seems to contain only at _b[0] the correct value of the current qp. 
Can you give me a hint on how to access all the variable values in the Kernel?

Thank you very much,
Philip

**Reply by GiudGiud** [reply]
is this a Kernel or an AuxKernel?
is this 2D or 3D?
Which quadrature are you specifying in the Executioner block? Or is it the default?

**Reply by philipJFpfahl** [reply]
Hey,
it is in an AuxKernel, in 2D.

```
[Executioner]
  type = Transient
  start_time = 0
  dt = 0.01
  num_steps = 2
  solve_type = 'PJFNK'
[]
```

Is there a place where I can read up on this?
Thank you,
Philip

**Reply by GiudGiud** [reply]
Hi Philip

So you actually wont have access to any other qp for a nodal variable. For an elemental variable you would, all the qp values are computed for the element. But for a nodal variable, the node is a the quadrature point, so there is only one value there. 

You could try performing this operation with an elemental variable, then projecting it to a nodal variable

Guillaume

**Reply by philipJFpfahl** [reply]
Hey,
Is it possible, that a kernel gets only handed the nodal position in the mesh and is then able to operate (however) on that node? Therefore it would not possible to access all the nodal values in each kernel. From a parallelization perspective that makes sense, but would make my life much more difficult.

Philip

**Reply by GiudGiud** [reply]
Yes. For nodes you only get to operate on that node, it's a single quadrature point loop basically

For elements you may work with all the quadrature points at once.

**Reply by philipJFpfahl** [reply]
I see. Is there a way to access the same Matrix and the same Variable, AuxVariable or vector at all individual nodes? 
Then I could do the Matrix multiplication that way.

**Reply by GiudGiud** [reply]
on nodes the classic way not that I know of.
@lindsayad how could we access multiple node values at the same time? We could re-init multiple times but that s not desirable

**Reply by lindsayad** [reply]
For a variable you could ask nodes for dof indices and then index the solution vector at those dof indices. I have only perused these threads but it seems like maybe a `GeneralUserObject` is the best object for doing these highly custom operations that require many dof indices at once

**Reply by GiudGiud** [reply]
@philipJFpfahl this operation is done on all the nodes of an element right?
I think you may want to have an element loop, access the nodes of the _current_elem, get the dof indices, and perform the matrix multiplication there, then set the dof indices in the result variable

**Reply by philipJFpfahl** [reply]
Hey,
Thanks for the opinion, I was looking at the GeneralUserObjects as well.
What do you mean by element? If that is a  Variable, AuxVariable, or vector ... then: 
Yes, that was my idea. I want to do a "normal" matrix multiplication where I put in an element and multiply it with a matrix and get an element
out of that. So 
$a = W \times b$ 
where $a, b \in N$ and $W \in N \times N$.
Do you have an example where something similar is done in MOOSE?
And should that be _corrent_node or element? I think I should stick to the nodal option since that makes the matrix generation easier for me.

Cheers,
Philip

**Reply by lindsayad** [reply]
@GiudGiud he said that he wants a mesh point to depend on all other mesh points. It sounds to me like this matrix should be sized roughly accordingly to the total number of nodes/dofs in the simulation and like this operation should happen on a global scale as opposed to on a per-element basis. But I could be wrong

**Reply by GiudGiud** [reply]
I m thinking ElementUserObject will be more suited than GeneralUserObject

I dont know of one

_current_node is for node-based loops. _current_elem for element-based loops. You will need to use an element loop

**Reply by philipJFpfahl** [reply]
So if I understand correctly the nodes are the mesh points and the elements are the points for the quadrature. 
I think "global" describes it good. What I want to do, in physics terms, is to weigh the importance of the neutrons produced by precursors in the next time step and add these as a neutron source to my neutron population on each mesh point.
I know the "importance" matrix. So I think the matrix should be the $(number of mesh points) ^2$.

**Reply by lindsayad** [reply]
Mesh nodes are a superset of element vertices. Elements may have mid-edge nodes, mid-face nodes, and mid-element nodes. Nodes store degrees of freedom for Lagrange variables. Elements store degrees of freedom for variable families like `MONOMIAL`. 

It sounds like you don't care about integration, in which case quadrature has no value to you. If you are using a Lagrange variable and all you care about is the values of the degrees of freedom in the global solution vector, then you will want to be doing node loops and `ElementUserObject` would not be suitable for you

**Reply by philipJFpfahl** [reply]
That sounds right. I want to use ```LAGRANGE``` type variables. The degrees of freedom would be equal to the number of mesh points I set up in the 
```
[Mesh]
[]
```
block?
How do I access the global solution vector in the ```ElementUserObject```?

**Reply by lindsayad** [reply]
A `NodalUserObject` could make sense for you. You could do global operations in `finalize`. The number of degrees of freedom will be equal to the number of mesh nodes if the order of the variable matches the order of the mesh geometry

**Reply by philipJFpfahl** [reply]
I think it does.  Is ```finalize``` called at the end of each time step or at the end of execution? I would need the $a$ at each time step.
Would a ```NodalUserObject``` give me access to the full $b$ vector?

**Reply by lindsayad** [reply]
User object execution is controlled by the `execute_on` parameter. You could tell it to `execute_on = timestep_end`. The at the end of each timestep, for a nodal user object, this is what happens:

1. `initialize` is called
2. Loop over all process-local nodes and call `execute` for each process-local node
3. `threadJoin` and `finalize` are called

You can access the solution vector in a user object if you ask for it from a `SystemBase` object with `solution()` (see the doxygen page [here](https://mooseframework.inl.gov/docs/doxygen/moose/classSystemBase.html#ad63de071e498c6f4582bded6abdc38b3))

**Reply by philipJFpfahl** [reply]
Okay, I see. You would not do it during ```execute``` since that would be called for each node?

Is an AuxVariable a ```SystemBase``` object? 
And how can I ask the Variable? $b.solution();$ certainly does not work. Maybe something more applied than the doxygen page could help ;).
Thank both of you for your effort.

Philip

**Reply by lindsayad** [reply]
> Okay, I see. You would not do it during execute since that would be called for each node?

I'm thinking about it more. Do you already know your matrix W? Or are you building it as you go? If you already know it, then you could be computing `a_j` at every node during `execute`. Alternatively if you wanted to compute the `a` vector all at once, then that might be better suited for a `GeneralUserObject`.

To get a solution vector ... if you want the nonlinear solution vector, then you can get it with `getCheckedPointerParam<SystemBase *>("_sys")->solution()` within your user object

**Reply by lindsayad** [reply]
The benefit of taking a by-node approach would be that you could be filling an auxiliary variable representation of `a` where `a` is a Lagrange auxiliary variable. In this case you would be using an `AuxKernel` and not a `NodalUserObject`.

If you use a `GeneralUserObject`, the question is what do you do with this big vector `a` after you compute it? How do you access that result in the places you need it?

**Reply by lindsayad** [reply]
If you used a (nodal) `AuxKernel`, then you can access the nonlinear solution vector with `_nl_sys.solution()`

**Reply by philipJFpfahl** [reply]
Hey,
Yes, I know the matrix before I start the moose application. It will be nuclear data from a Montecarlo run. 

If I can access the complete ```b``` vector during each loop in the ```AuxKernel``` that would be perfect. That was the goal of my original question to find the data type, that I could access during the run. 

Yes having an ```AuxVariable``` as an output seems also like the right choice since I want to do stuff with that later.

Also I can implement the matrix as a ```Eigen``` matrix. 

Is the the nonlinear solution my coupledValue or is it the total solution vector?

Philip

**Reply by lindsayad** [reply]
`_nl_sys.solution()` is the "total" solution vector. Note that it includes the solution entries for all variables in the nonlinear system. If you have more than one variable in the nonlinear system, you will need to do some work to determine what the dof indices you want to index at specific to your variable. You could do this during `initialSetup` where you could loop over `_mesh.getActiveNodeRange()` and query the nodes for dof indices of the nonlinear variable you care about and store it on a data member in your aux kernel

**Reply by philipJFpfahl** [reply]
Would I get the solution to my coupled variable with ```getCheckedPointerParam<SystemBase *>("_sys")->solution()```?
Is ```initialSetup``` called at each time step or only once at the execution?

Is all the data stored in one solution vector and do I have to figure out where in that vector the desired variable is stored? (Using ```_mesh.getActiveNodeRange()```)

Thanks again this is a great help!

**Reply by lindsayad** [reply]
You'd want
```
_b_number(coupled("b")),
_serial_solution(_nl_sys.serializedSolution())
```
and then assuming you have a `std::vector<dof_id_type> _b_indices;` data member:
```c++
void
PhilipsAuxKernel::determineDofIndices()
{
  _b_indices.clear();
  _b_indices.reserve(_mesh.getMesh().n_nodes());

  for (auto * const node : _mesh.getMesh().local_node_ptr_range())
  {
    if (!node->active())
      continue;

    _b_indices.push_back(node->dof_number(0, _b_number, 0));
  }

  _communicator.allgather(_b_indices, false);
  _b_indices.shrink_to_fit();
}

void
PhilipsAuxKernel::initialSetup()
{
  determineDofIndices();
}

void
PhilipsAuxKernel::meshChanged()
{
  determineDofIndices();
}
```
You will now have all your variable's dof indices, and you can use those to index into `_serial_solution`

**Reply by philipJFpfahl** [reply]
oh wow, that's more than perfect!
Would you load the Matrix as an Eigen::Matrix or would you use a Moose data type?
Maybe a moose vector is better when the mesh is more complex.

**Reply by lindsayad** [reply]
Since it sounds like it will be a dense multiplication, maybe Eigen is a good choice for that

**Reply by philipJFpfahl** [reply]
Hey, 
I struggle to implement that code bit above. 

firstly the line
```_b_indices.push_node->dof_number(0, _b_number, 0);```
only compiles with
 ```vector::push_node```
 and the
```dof_number(0, _b_number, 0)``` seems not to work. It seems that ```dof_number()``` expects 3 int type inputs. 

Secondly, how do I index the ```_nl_sys.serializedSolution()```?

**Reply by lindsayad** [reply]
Sometimes I type too fast. I've edited my code snippet above. It should have read `_b_indices.push_back(node->dof_number(0, _b_number, 0));`.

Index the solution vector like this:
`_serial_solution(0)`
but using the indices stored in `_b_indices`.

**Reply by philipJFpfahl** [reply]
Hey,
It works fine. For completeness in the .h file I have:
```  
  void initialSetup();
  void determineDofIndices();
  void meshChanged();

  const unsigned int _b_number;
  std::vector<dof_id_type> _b_indices;
  libMesh::NumericVector<double> & _serial_solution;
```

The next problem with this task is that I have to implement the Matrix in a way, that I can weight the right ```_serial_solution``` with the right entry in the matrix. I already figured out, that libMesh has a different way of indexing, than I expected. 

For each node I could load a vector, and do the matrix operation with that.

```
Real
PhilipsAuxKernel::computeValue()
{  

for(int i = 0; i<_b_indices.size(); i++ ){
  _weight_sum +=  _weight_vector[i] * _serial_solution(_b_indices[i]); 
};
    
return _weight_sum ;
}
```
I wonder the best way to load the ```_weight_vector``` so that the indexing is right. 
Can you help me with that too or is that a libMesh problem?


Thank you
Philip

**Comment by philipJFpfahl** [comment]
Hey, 
I am still struggling with the implementation of the weight matrix. It seems like the order in which MOOSE addresses each node is not row by row. 
What I mean, is that if I have a $2 \times 2$ grid I expect the indexing to be something like this:
```
0 1  
2 3
```
resulting in a solution vector indexed like this ``` (0, 1, 2, 3) ```. For this, I would know, how to load the weight matrix.
But the indexing is more like this:
```
0 1
3 2
```
resulting in ``` (0, 1, 3, 2) ``` as the indices for the returned solution vector.
Is there a way, to load in (maybe a material) and have it be in the same order as the MOOSE solution vector?
what I mean is that for a matrix with indexes like that:
```
00 01 02 03
10 11 12 13 
20 21 22 23 
30 31 32 33 
...
```
I would  need it to be transformed into a matrix like this: 
```
00 01 03 02
10 11 13 12 
30 31 33 32 
20 21 23 22 
```
Note that for $W_{j,i}$ both the $j$ and the $i$ index are "out of order".
I don't have a better description. Maybe there is also a way to get the "real" position of the node somehow. All ideas are welcome.

Thank you,
Philip

**Reply by GiudGiud** [reply]
if you dump the material into an auxvariable it will be ordered similarly as other variables. 
You may use Array aux variables for this

**Reply by philipJFpfahl** [reply]
Hey,
I think that is the right path. The AuxVariabes are stored in the same "nodal" order as the Variables. I also found out, that  ```allow_renumbering = false``` orders the nodes in a better way. (As I would do it by hand)
Now the problem is the following:
I still need access to each entry in the AuxVariabes vector. So how do I do the same you showed me for Variables with AuxVariables?
All the best,
Philip

**Reply by lindsayad** [reply]
How do you compute W? I feel like that’s a very important question because maybe it would inform my question of how you would get W correctly sorted/stored into an array aux variable representation

**Reply by lindsayad** [reply]
By the way somewhere in this discussion I think you asked about accessing positions. If you are in a nodal aux kernel, you have a _current_node data member which is a pointer to the current Node, which derives from Point. So dereferencing the pointer, you will have something that behaves exactly like a Point. So for instance you can access the x-coordinate like this:

(*_current_node)(0)

**Reply by philipJFpfahl** [reply]
$W$ is the tensor product of two vectors $c \otimes d$ with the same length as $a$. 

In physical terms, the two vectors are scaled versions of the power shape of the reactor. This matrix is not really sparse. My idea was only to store the two vectors and then calculate the entries of $W$ in each $for$ loop. This will work fast since the problem seems memory bound. (Please correct me if I am wrong)

I load the AuxVariables with the ```PropertyReadFile``` Object. 

Philip

**Reply by philipJFpfahl** [reply]
> (*_current_node)(0)
Thats a good tip. Thank you, I was just confused in which way MOOSE indexes the different points. That is also how I intend to find the current node id ``` _current_node -> id()```.
As I said I am just getting into MOOSE.

**Reply by lindsayad** [reply]
Sorry for the delay @philipJFpfahl. I am currently on leave but will be back to work next Monday at which time I should be able to help more

**Reply by philipJFpfahl** [reply]
Oh, thank you for checking up. 
I figured out that I can find the aux solution vector with ```_aux_sys.serializedSolution()``` but the ```coupled()``` method does not return the correct index for auxVariables. The rest should be working fine. 

All the best and have a nice time off.

**Reply by philipJFpfahl** [reply]
Hey,
have you found a solution to find out the right integer for the position in the aux solution vector?

all the best
Philip

**Reply by lindsayad** [reply]
I forget where we were at in the conversation. You can get the correct dof index if you use the [DofObject::dof_number API](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1DofObject.html#a4e8bf039eb9c13b7eb3aaf96afc8c8b8). So assuming `_aux_var_number = coupled("blah_blah")`:
```c++
const auto soln_vector_index = _current_node->dof_number(_aux_sys.number(), _aux_var_number, 0);
```
Here I am assuming you are in a nodal aux kernel (I think that's where we arrived, but correct me if I'm wrong)

**Reply by philipJFpfahl** [reply]
Hey,

you remember correctly. 
The problem is for ```_aux_var_number = coupled("blah_blah")``` I do not get the right index to find it in  ```_aux_sys```.
The index should be a small int, but it is a really big int (4294967294).

Philip

**Reply by GiudGiud** [reply]
is blah_blah the name of the variable? 
you want to be:
- passing a coupled var as a parameter params.addCoupledVar("name_of_param" <- this is not the name of the variable, it s the parameter
- in input: name_of_param = variable_name
- in code, initialize the index in the constructor as _var_index(coupled("name_of_param"))

it should work so if it doesnt we ll need your code & input to find out

**Reply by philipJFpfahl** [reply]
Yes it works fine with a variable it does not work with an aux_variable. 

Lets say I want to calculate at each node ``` for all mesh points:  a +=  b[i] * C[i]```

then I would use this as my h file:
```
#pragma once

#include "AuxKernel.h"


class scalar_multiplication : public AuxKernel
{
public:
  static InputParameters validParams();

  scalar_multiplication(const InputParameters & parameters);

protected:
  virtual Real computeValue() override;
  void initialSetup();
  void determineDofIndices();
  void meshChanged();
  
  const unsigned int _c_number;
  std::vector<dof_id_type> _c_indices;
  
  const unsigned int _b_number;
  std::vector<dof_id_type> _b_indices;
   
  //contains solution vector
  libMesh::NumericVector<double> & _c_serial_solution;
  libMesh::NumericVector<double> & _b_serial_solution;
  double _sum;
};
```
with the scalar_multiplication.C file
```
#include "scalar_multiplication.h"
#include "SystemBase.h"

#include "AuxiliarySystem.h"

registerMooseObject("MyApp", scalar_multiplication);
InputParameters
scalar_multiplication::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription("Returns b*c");

  params.addRequiredCoupledVar("c", "variable c");
  params.addRequiredCoupledVar("b", "aux variable b");
  
  return params;
}

scalar_multiplication::scalar_multiplication(const InputParameters & parameters)
  : AuxKernel(parameters),
  
  _c_number(coupled("c")),
  _b_number(coupled("b")),

  _c_serial_solution(_nl_sys.serializedSolution()),
  _b_serial_solution(_aux_sys.serializedSolution())
{
}



void
scalar_multiplication::determineDofIndices()
{
  _c_indices.clear();
  _c_indices.reserve(_mesh.getMesh().n_nodes());
  _b_indices.clear();
  _b_indices.reserve(_mesh.getMesh().n_nodes());

  for (auto * const node : _mesh.getMesh().local_node_ptr_range())
  {
    if (!node->active())
      continue;
    //This one works fine
    _c_indices.push_back(node->dof_number(0, _c_number, 0));
   // This one does not work _b_number should be 0 if b is initialized first in the input file.
    _b_indices.push_back(node->dof_number(_aux_sys.number(), _b_number, 0));
   //    _b_indices.push_back(node->dof_number(_aux_sys.number(), 0, 0));
  }

  _communicator.allgather(_c_indices, false);
  _c_indices.shrink_to_fit();
  _communicator.allgather(_b_indices, false);
  _b_indices.shrink_to_fit();
}

void
scalar_multiplication::initialSetup()
{
  determineDofIndices();
}

void
scalar_multiplication::meshChanged()
{
  determineDofIndices();
}
Real
scalar_multiplication::computeValue()
{  
   
    _sum = 0;
    if (isNodal()){
        for(unsigned int i = 0; i<_c_indices.size(); i++){
          _sum += _b_serial_solution(_b_indices[i]) *  _c_serial_solution(_c_indices[i]);
        };
    }
    else
        std::cout << "The AuxKernel is elemental" << "\n";
    return _sum;
}

```
with this input file 
```

[Mesh]
  allow_renumbering = false
  type = GeneratedMesh
  dim = 2 
  xmin = 0 
  xmax = 2
  nx = 2
  ymin = 0
  ymax = 2
  ny = 2
[]


[AuxVariables]
  [b]
    order = FIRST
    family = LAGRANGE
  []
  [a]
    order = FIRST
    family = LAGRANGE
  []
[]

[Variables]
  [c]
    family = LAGRANGE
  []
[]


[Kernels]
  [c_time]
  type = TimeDerivative
  variable = c
  []
[]


[AuxKernels]
  [sclar_multiplication]
    type = scalar_multiplication
    variable = a
    b = "b"
    c = "c"
    execute_on = 'TIMESTEP_END'
  []
[]

[Functions]
  [a_func]
    type = ParsedFunction
    value = 0
  []
  [b_func]
    type = ParsedFunction
    value = 1
  []
  [c_func]
    type = ParsedFunction
    value = 1 
  []
[]

[ICs]
  [Initial_a]
    type = FunctionIC
    variable = 'a'
    function = a_func
  []
  [IC_b]
    type = FunctionIC
    variable = 'b'
    function = 'b_func'
  []
  [Initial_c]
    type = FunctionIC
    variable = 'c'
    function = c_func 
  []
[]

[Executioner]
  type = Transient
  start_time = 0
  dt = 1 
  num_steps = 2 
  solve_type = 'PJFNK'
[]

[Outputs]
  exodus = true
[]

```
This all works fine when I manually enter the right index for ```_b_number``` in ```b_indices.push_back(node->dof_number(_aux_sys.number(), _b_number, 0));```. This option is commented out in the .C file above.

a is then  0 at the beginning and then 9 at each node (since I am suming 1*1 nine times). 

So maybe for aux variables ```  _b_number(coupled("b"))``` does not work properly for aux_variables?


Sorry for the long post. 
Philip

**Reply by lindsayad** [reply]
Oh my word, I have no idea why the code is this way. This is really stupid
```c++
unsigned int
Coupleable::coupled(const std::string & var_name, unsigned int comp) const
{
  const auto * var = getFieldVar(var_name, comp);
  if (!var)
  {
    mooseAssert(_optional_var_index.find(var_name) != _optional_var_index.end(),
                "optional var index for " << var_name << " does not exist!");
    // make sure we don't try to access default var ids that were not provided
    checkComponent(_obj, comp, _optional_var_index.at(var_name).size(), var_name);
    return _optional_var_index.at(var_name)[comp];
  }
  checkFuncType(var_name, VarType::Ignore, FuncAge::Curr);

  switch (var->kind())
  {
    case Moose::VAR_NONLINEAR:
      return var->number();
    case Moose::VAR_AUXILIARY:
      return std::numeric_limits<unsigned int>::max() - var->number();
    default:
      mooseError(_c_name, ": Unknown variable kind. Corrupted binary?");
  }
}
```

**Reply by lindsayad** [reply]
well there is a reason for it. This is the original PR that added this: https://github.com/idaholab/moose/issues/1504. What I would do @philipJFpfahl is change to do this:
```c++
  _c_number(coupled("c")),
  _b_number(getVar("b", 0)->number()),
```
You could also do that with 'c' if you think you might every want to use an aux variable for that parameter

**Reply by philipJFpfahl** [reply]
Hey,

It feels like that reason could have been solved differently. But now it is working fine. 

Thanks for all the effort. 
Philip
